asyncpg/connection.py

```diff
@@ -586,13 +624,38 @@ def _drop_global_statement_cache(self):
             # for discussion.
             #
             self._drop_global_statement_cache()
+            if self._protocol.is_in_transaction() or not retry:
+                raise
+            else:
+                result = await self._do_execute(
+                    query, executor, retry=False)
+
+        return result
+
+    async def _do_execute_with_timeout(self, query, executor, timeout,
```

Comment:
Even though I suggested to split `_do_execute` in two methods myself, I now think we'd be better off to merge them in one :) The actual timeouts arithmetic happens just in two places, so we can just add two `if` statements to `_do_execute`.